### DISCUSSION TITLE
Try the new Generate Commit Message (Experimental)

### DISCUSSION THREAD

**Discussion by SpaceMerlin33** [gitkraken/vscode-gitlens#2581] [Labs]
We’ve introduced a new experimental feature to assist with writing commit messages by leveraging OpenAI. 

![Generate Commit OpenAI](https://user-images.githubusercontent.com/110862234/225744135-4d3f3514-5532-442a-a075-9c6ec653ffb1.gif)

To start, stage some changes you want to commit and generate a commit message for and then run the “Generate Commit Message (Experimental)” command from the Command Palette. From there you will be guided through the process of accepting to send the diff of your staged changes to OpenAI and the entry of your OpenAI key. Once completed, the generated commit message will be entered into the commit input box on the Source Control sidebar. You can also enter additional context about your changes in the commit box and those will also be sent to help generate a better message.

Additionally, you can customize the `gitlens.experimental.generateCommitMessagePrompt` setting to control the prompt used to structure and format the generated commit message. You could have it structure commit messages in the conventional commit style, or you can have fun with it and make your commit messages in the style of a pirate, Shakespeare, etc.

This is a very early peek into features and ideas that can leverage AI – with more planned in the near future. We're excited to bring this experimental feature to you and are ready to answer any questions and make improvements where needed.

**We'd love to hear your feedback! 🙏**

**Comment by gllsrssx** [comment]
works well but i cannot change openai api key

**Reply by jkelroy** [reply]
I see you answered this [here](https://github.com/gitkraken/vscode-gitlens/issues/2665#issuecomment-1540538159), but will include the answer to this thread as well: run the command `GitLens: Reset Stored OpenAI Key` from the Command Palette.

**Comment by PeterNjeim** [comment]
This is great, however I don't want to pay for an OpenAI key. Maybe in the future you can use open source models like [LAION OpenAssistant](https://open-assistant.io), [HuggingChat](https://huggingface.co/chat), or the new [Falcon-Chat](https://huggingface.co/spaces/HuggingFaceH4/falcon-chat) (currently not tuned for code but it is the [best open source LLM](https://huggingface.co/spaces/HuggingFaceH4/open_llm_leaderboard) currently)

**Reply by jkelroy** [reply]
GitLens currently supports OpenAI and Anthropic. I created a feature request to support additional models here. https://github.com/gitkraken/vscode-gitlens/issues/2743#issue-1759442335

**Comment by lestephane** [comment]
Can it be coerced into generating conventional commits? https://www.conventionalcommits.org

**Reply by jkelroy** [reply]
You can use the [setting](https://help.gitkraken.com/gitlens/gitlens-settings/#misc-settings) `gitlens.experimental.generateCommitMessagePrompt` to prompt for the response in the desired format.

**Reply by LukasPolak** [reply]
I have the following prompt, but I'm not getting the conventional commits output.
Do you have any ideas on how to fix the prompt @jkelroy? 🙏 

```
Generate a commit message using the Conventional Commits format. Examples: ['feat: Add new feature to the project', 'fix: Fix a bug in the project', 'chore: Update build configuration or task', 'docs: Update project documentation', 'style: Update code formatting or style', 'refactor: Refactor existing code', 'test: Add or update tests', 'perf: Improve performance of the project', 'ci: Update continuous integration configuration', 'build: Make changes related to the build process', 'revert: Revert a previous commit']
```

Output is following:

```
Update VSCode settings.json with eslint and prettier configurations
```

**Reply by jkelroy** [reply]
Do you have the string wrapped in quotes in your settings.json? To check this, open the command palette (ctrl/command + shift + P) and search _Preferences: Open User Settings (JSON)_. The entry should look like this - if you have more settings before or after, make sure to include a `,` after each entry:

```
{
    "gitlens.experimental.generateCommitMessagePrompt": "Generate a commit message using the Conventional Commits format. Examples: ['feat: Add new feature to the project', 'fix: Fix a bug in the project', 'chore: Update build configuration or task', 'docs: Update project documentation', 'style: Update code formatting or style', 'refactor: Refactor existing code', 'test: Add or update tests', 'perf: Improve performance of the project', 'ci: Update continuous integration configuration', 'build: Make changes related to the build process', 'revert: Revert a previous commit']"
}
```

**Reply by LukasPolak** [reply]
```json
{
  "gitlens.experimental.generateCommitMessagePrompt": "Generate a commit message using the Conventional Commits format. Examples: ['feat: Add new feature to the project', 'fix: Fix a bug in the project', 'chore: Update build configuration or task', 'docs: Update project documentation', 'style: Update code formatting or style', 'refactor: Refactor existing code', 'test: Add or update tests', 'perf: Improve performance of the project', 'ci: Update continuous integration configuration', 'build: Make changes related to the build process', 'revert: Revert a previous commit']"
}
```

I verified the validity of the JSON, and it is valid. Additionally, the other settings are functioning properly.

**Reply by jkelroy** [reply]
I see, so the commit message is being generated but seemingly just not getting the desired output based on the prompt. What AI provider and model are you using?

**Reply by LukasPolak** [reply]
Yes, that is right

```
  "gitlens.ai.experimental.provider": "openai",
  "gitlens.ai.experimental.openai.model": "gpt-3.5-turbo",
```

**Reply by LukasPolak** [reply]
I realized I used GitHub Copilot instead of GitLens to generate commit messages.
Apologies for the inconvenience and thank you for the prompt responses!

**Comment by mikernet** [comment]
I'm not sure if github copilot has integration points that can be leveraged so it can be used as a commit message provider in GK or not. Are there any plans to support this scenario? Would be nice, given that we already have subs for that.

**Reply by jkelroy** [reply]
There is an existing feature request to support additional AI models: https://github.com/gitkraken/vscode-gitlens/issues/2743

**Comment by str** [comment]
We use git-flow where every feature branch is feature/ticket-number-small-description.

We use the internal standard to have as the commit header "ticket-number Small Description", like "https://github.com/gitkraken/vscode-gitlens/discussions/4530 Commit Composer View". Is there a way to define that rule for the suggestion to be used,? Somewhere (like in .github/copilot-instructions.md)

Thank you.
